File: CS/SuppliersExample/Form1.cs
#region #Namespaces
using System;
using System.Windows.Forms;
using DevExpress.Spreadsheet;
using SuppliersExample.NWindDataSetTableAdapters;
// ...
#endregion #Namespaces

namespace SuppliersExample
{
    public partial class Form1 : DevExpress.XtraBars.Ribbon.RibbonForm
    {
        bool applyChangesOnRowsRemoved = false;
        #region #BindToData
        NWindDataSet dataSet;
        SuppliersTableAdapter adapter;

        public Form1() {
            InitializeComponent();
            dataSet = new NWindDataSet();
            adapter = new SuppliersTableAdapter();
        }

        private void Form1_Load(object sender, EventArgs e) {
            // Populate the "Suppliers" data table with data.
            adapter.Fill(dataSet.Suppliers);
            IWorkbook workbook = spreadsheetControl1.Document;
            // Load the template document into the SpreadsheetControl.
            workbook.LoadDocument("Suppliers_template.xlsx");
            Worksheet sheet = workbook.Worksheets[0];
            // Load data from the "Suppliers" data table into the worksheet starting from the cell "B12".
            sheet.DataBindings.BindToDataSource(dataSet.Suppliers, 11, 1);
        }
        #endregion #BindToData

        #region #UpdateData
        private void spreadsheetControl1_MouseClick(object sender, MouseEventArgs e) {
            Cell cell = spreadsheetControl1.GetCellFromPoint(e.Location);
            if (cell == null)
                return;
            Worksheet sheet = spreadsheetControl1.ActiveWorksheet;
            string cellReference = cell.GetReferenceA1();
            // If the "Save" cell is clicked in the data entry form, 
            // add a row containing the entered values to the database table.
            if (cellReference == "I4") {
                AddRow(sheet);
                HideDataEntryForm(sheet);
                ApplyChanges();
            }
            // If the "Cancel" cell is clicked in the data entry form, 
            // cancel adding new data and hide the data entry form.
            else if (cellReference == "I6") {
                HideDataEntryForm(sheet);
            }
        }

        void AddRow(Worksheet sheet) {
            try {
                // Append a new row to the "Suppliers" data table.
                dataSet.Suppliers.AddSuppliersRow(
                    sheet["C4"].Value.TextValue, sheet["C6"].Value.TextValue, sheet["C8"].Value.TextValue,
                    sheet["E4"].Value.TextValue, sheet["E6"].Value.TextValue, sheet["E8"].Value.TextValue,
                    sheet.Cells["G4"].DisplayText, sheet.Cells["G6"].DisplayText);
            }
            catch (Exception ex) {
                string message = string.Format("Cannot add a row to a database table.\n{0}", ex.Message);
                MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        void HideDataEntryForm(Worksheet sheet) {
            Range range = sheet.Range.Parse("C4,C6,C8,E4,E6,E8,G4,G6");
            range.ClearContents();
            sheet.Rows.Hide(2, 9);
        }

        void ApplyChanges() {
            try {
                // Send the updated data back to the database.
                adapter.Update(dataSet.Suppliers);
            }
            catch (Exception ex) {
                string message = string.Format("Cannot update data in a database table.\n{0}", ex.Message);
                MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        private void spreadsheetControl1_RowsRemoving(object sender, RowsChangingEventArgs e) {
            Worksheet sheet = spreadsheetControl1.ActiveWorksheet;
            Range rowRange = sheet.Range.FromLTRB(0, e.StartIndex, 16383, e.StartIndex + e.Count - 1);
            Range boundRange = sheet.DataBindings[0].Range;
            // If the rows to be removed belong to the data-bound range,
            // display a dialog requesting the user to confirm the deletion of records. 
            if (boundRange.IsIntersecting(rowRange)) {
                DialogResult result = MessageBox.Show("Want to delete the selected supplier(s)?", "Delete", 
                    MessageBoxButtons.YesNo, MessageBoxIcon.Question);
                applyChangesOnRowsRemoved = result == DialogResult.Yes;
                e.Cancel = result == DialogResult.No;
                return;
            }
        }

        private void spreadsheetControl1_RowsRemoved(object sender, RowsChangedEventArgs e) {
            if (applyChangesOnRowsRemoved) {
                applyChangesOnRowsRemoved = false;
                // Update data in the database.
                ApplyChanges();
            }
        }

        private void buttonAddRecord_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            CloseInplaceEditor();
            Worksheet sheet = spreadsheetControl1.ActiveWorksheet;
            // Display the data entry form on the worksheet to add a new record to the "Suppliers" data table.
            if (!sheet.Rows[4].Visible)
                sheet.Rows.Unhide(2, 9);
            spreadsheetControl1.SelectedCell = sheet["C4"];
        }

        private void buttonRemoveRecord_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            CloseInplaceEditor();
            Worksheet sheet = spreadsheetControl1.ActiveWorksheet;
            Range selectedRange = spreadsheetControl1.Selection;
            Range boundRange = sheet.DataBindings[0].Range;
            // Verify that the selected cell range belongs to the data-bound range.
            if (!boundRange.IsIntersecting(selectedRange) || selectedRange.TopRowIndex < boundRange.TopRowIndex) {
                MessageBox.Show("Select a record first!", "Remove Record", MessageBoxButtons.OK, MessageBoxIcon.Error);
                return;
            } 
            // Remove the topmost row of the selected cell range.
            sheet.Rows.Remove(selectedRange.TopRowIndex);
        }

        private void buttonApplyChanges_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            CloseInplaceEditor();
            // Update data in the database.
            ApplyChanges();
        }

        private void buttonCancelChanges_ItemClick(object sender, DevExpress.XtraBars.ItemClickEventArgs e) {
            // Close the cell in-place editor if it's currently active. 
            CloseInplaceEditor();
            // Load the latest saved data into the "Suppliers" data table.
            adapter.Fill(dataSet.Suppliers);
        }

        void CloseInplaceEditor() {
            if (spreadsheetControl1.IsCellEditorActive)
                spreadsheetControl1.CloseCellEditor(DevExpress.XtraSpreadsheet.CellEditorEnterValueMode.Default);
        }
        #endregion #UpdateData
    }
}

File: VB/SuppliersExample/Form1.vb
#Region "#Namespaces"
Imports System
Imports System.Windows.Forms
Imports DevExpress.Spreadsheet
Imports SuppliersExample.NWindDataSetTableAdapters

' ...
#End Region  ' #Namespaces
Namespace SuppliersExample

    Public Partial Class Form1
        Inherits DevExpress.XtraBars.Ribbon.RibbonForm

        Private applyChangesOnRowsRemoved As Boolean = False

#Region "#BindToData"
        Private dataSet As NWindDataSet

        Private adapter As SuppliersTableAdapter

        Public Sub New()
            InitializeComponent()
            dataSet = New NWindDataSet()
            adapter = New SuppliersTableAdapter()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            ' Populate the "Suppliers" data table with data.
            adapter.Fill(dataSet.Suppliers)
            Dim workbook As IWorkbook = spreadsheetControl1.Document
            ' Load the template document into the SpreadsheetControl.
            workbook.LoadDocument("Suppliers_template.xlsx")
            Dim sheet As Worksheet = workbook.Worksheets(0)
            ' Load data from the "Suppliers" data table into the worksheet starting from the cell "B12".
            sheet.DataBindings.BindToDataSource(dataSet.Suppliers, 11, 1)
        End Sub

#End Region  ' #BindToData
#Region "#UpdateData"
        Private Sub spreadsheetControl1_MouseClick(ByVal sender As Object, ByVal e As MouseEventArgs)
            Dim cell As Cell = spreadsheetControl1.GetCellFromPoint(e.Location)
            If cell Is Nothing Then Return
            Dim sheet As Worksheet = spreadsheetControl1.ActiveWorksheet
            Dim cellReference As String = cell.GetReferenceA1()
            ' If the "Save" cell is clicked in the data entry form, 
            ' add a row containing the entered values to the database table.
            If Equals(cellReference, "I4") Then
                AddRow(sheet)
                HideDataEntryForm(sheet)
                ApplyChanges()
            ' If the "Cancel" cell is clicked in the data entry form, 
            ' cancel adding new data and hide the data entry form.
            ElseIf Equals(cellReference, "I6") Then
                HideDataEntryForm(sheet)
            End If
        End Sub

        Private Sub AddRow(ByVal sheet As Worksheet)
            Try
                ' Append a new row to the "Suppliers" data table.
                dataSet.Suppliers.AddSuppliersRow(sheet("C4").Value.TextValue, sheet("C6").Value.TextValue, sheet("C8").Value.TextValue, sheet("E4").Value.TextValue, sheet("E6").Value.TextValue, sheet("E8").Value.TextValue, sheet.Cells("G4").DisplayText, sheet.Cells("G6").DisplayText)
            Catch ex As Exception
                Dim message As String = String.Format("Cannot add a row to a database table." & Microsoft.VisualBasic.Constants.vbLf & "{0}", ex.Message)
                MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End Sub

        Private Sub HideDataEntryForm(ByVal sheet As Worksheet)
            Dim range As CellRange = sheet.Range.Parse("C4,C6,C8,E4,E6,E8,G4,G6")
            range.ClearContents()
            sheet.Rows.Hide(2, 9)
        End Sub

        Private Sub ApplyChanges()
            Try
                ' Send the updated data back to the database.
                adapter.Update(dataSet.Suppliers)
            Catch ex As Exception
                Dim message As String = String.Format("Cannot update data in a database table." & Microsoft.VisualBasic.Constants.vbLf & "{0}", ex.Message)
                MessageBox.Show(message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error)
            End Try
        End Sub

        Private Sub spreadsheetControl1_RowsRemoving(ByVal sender As Object, ByVal e As RowsChangingEventArgs)
            Dim sheet As Worksheet = spreadsheetControl1.ActiveWorksheet
            Dim rowRange As CellRange = sheet.Range.FromLTRB(0, e.StartIndex, 16383, e.StartIndex + e.Count - 1)
            Dim boundRange As CellRange = sheet.DataBindings(0).Range
            ' If the rows to be removed belong to the data-bound range,
            ' display a dialog requesting the user to confirm the deletion of records. 
            If boundRange.IsIntersecting(rowRange) Then
                Dim result As DialogResult = MessageBox.Show("Want to delete the selected supplier(s)?", "Delete", MessageBoxButtons.YesNo, MessageBoxIcon.Question)
                applyChangesOnRowsRemoved = result = DialogResult.Yes
                e.Cancel = result = DialogResult.No
                Return
            End If
        End Sub

        Private Sub spreadsheetControl1_RowsRemoved(ByVal sender As Object, ByVal e As RowsChangedEventArgs)
            If applyChangesOnRowsRemoved Then
                applyChangesOnRowsRemoved = False
                ' Update data in the database.
                ApplyChanges()
            End If
        End Sub

        Private Sub buttonAddRecord_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            CloseInplaceEditor()
            Dim sheet As Worksheet = spreadsheetControl1.ActiveWorksheet
            ' Display the data entry form on the worksheet to add a new record to the "Suppliers" data table.
            If Not sheet.Rows(4).Visible Then sheet.Rows.Unhide(2, 9)
            spreadsheetControl1.SelectedCell = sheet("C4")
        End Sub

        Private Sub buttonRemoveRecord_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            CloseInplaceEditor()
            Dim sheet As Worksheet = spreadsheetControl1.ActiveWorksheet
            Dim selectedRange As CellRange = spreadsheetControl1.Selection
            Dim boundRange As CellRange = sheet.DataBindings(0).Range
            ' Verify that the selected cell range belongs to the data-bound range.
            If Not boundRange.IsIntersecting(selectedRange) OrElse selectedRange.TopRowIndex < boundRange.TopRowIndex Then
                MessageBox.Show("Select a record first!", "Remove Record", MessageBoxButtons.OK, MessageBoxIcon.Error)
                Return
            End If

            ' Remove the topmost row of the selected cell range.
            sheet.Rows.Remove(selectedRange.TopRowIndex)
        End Sub

        Private Sub buttonApplyChanges_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            CloseInplaceEditor()
            ' Update data in the database.
            ApplyChanges()
        End Sub

        Private Sub buttonCancelChanges_ItemClick(ByVal sender As Object, ByVal e As DevExpress.XtraBars.ItemClickEventArgs)
            ' Close the cell in-place editor if it's currently active. 
            CloseInplaceEditor()
            ' Load the latest saved data into the "Suppliers" data table.
            adapter.Fill(dataSet.Suppliers)
        End Sub

        Private Sub CloseInplaceEditor()
            If spreadsheetControl1.IsCellEditorActive Then spreadsheetControl1.CloseCellEditor(DevExpress.XtraSpreadsheet.CellEditorEnterValueMode.Default)
        End Sub
#End Region  ' #UpdateData
    End Class
End Namespace
